src/xtea.cpp

```diff
@@ -0,0 +1,213 @@
+    }
+}
+
+template<size_t BLOCK_SIZE>
+void XTEA_decrypt(uint8_t data[BLOCK_SIZE * 8], const key& k)
+{
+    alignas(16) uint32_t left[BLOCK_SIZE], right[BLOCK_SIZE];
+    for (auto i = 0u, j = 0u; i < BLOCK_SIZE; i += 1u, j += 8u) {
+        left[i] = data[j] | data[j+1] << 8u | data[j+2] << 16u | data[j+3] << 24u;
+        right[i] = data[j+4] | data[j+5] << 8u | data[j+6] << 16u | data[j+7] << 24u;
+    }
+
+    uint32_t sum = delta * 32;
+    for (auto i = 0u; i < 32; ++i) {
+        for (auto j = 0u; j < BLOCK_SIZE; ++j) {
+            right[j] -= (((left[j] << 4) ^ (left[j] >> 5)) + left[j]) ^ (sum + k[(sum >> 11) & 3]);
+        }
+        sum -= delta;
+        for (auto j = 0u; j < BLOCK_SIZE; ++j) {
+            left[j] -= (((right[j] << 4) ^ (right[j] >> 5)) + right[j]) ^ (sum + k[(sum) & 3]);
+        }
+    }
+
+    for (auto i = 0u, j = 0u; i < BLOCK_SIZE; i += 1u, j += 8u) {
+        data[j] = static_cast<uint8_t>(left[i]);
+        data[j+1] = static_cast<uint8_t>(left[i] >> 8u);
+        data[j+2] = static_cast<uint8_t>(left[i] >> 16u);
+        data[j+3] = static_cast<uint8_t>(left[i] >> 24u);
+        data[j+4] = static_cast<uint8_t>(right[i]);
+        data[j+5] = static_cast<uint8_t>(right[i] >> 8u);
+        data[j+6] = static_cast<uint8_t>(right[i] >> 16u);
+        data[j+7] = static_cast<uint8_t>(right[i] >> 24u);
+    }
+}
+
+} // anonymous namespace
+
+void encrypt(uint8_t* data, size_t length, const key& k)
+{
+    assert(length % 8 == 0);
+
+#ifdef __AVX512F__
+    {
+        constexpr auto step = 128u * 8u;
+        const auto blocks = (length & ~(step - 1));
+        for (auto i = 0u; i < blocks; i += step) {
+            XTEA_encrypt<128>(data + i, k);
+        }
+        data += blocks;
+        length -= blocks;
+    }
+#endif // __AVX512F__
+
+#ifdef __AVX__
+    {
+        constexpr auto step = 32u * 8u;
+        const auto blocks = (length & ~(step - 1));
+        for (auto i = 0u; i < blocks; i += step) {
+            XTEA_encrypt<32>(data + i, k);
+        }
+        data += blocks;
+        length -= blocks;
+    }
+#endif // __AVX__
+
+#if defined(__SSE__) || defined(__ARM_FEATURE_SIMD32)
+    {
+        constexpr auto step = 8u * 8u;
+        const auto blocks = (length & ~(step - 1));
+        for (auto i = 0u; i < blocks; i += step) {
+            XTEA_encrypt<8>(data + i, k);
+        }
+        data += blocks;
+        length -= blocks;
+    }
+#endif // __SSE__ || __ARM_FEATURE_SIMD32
+
+#ifdef __x86_64__
+    {
+        constexpr auto step = 2u * 8u;
+        const auto blocks = (length & ~(step - 1));
+        for (auto i = 0u; i < blocks; i += step) {
+            XTEA_encrypt<2>(data + i, k);
+        }
+        data += blocks;
+        length -= blocks;
+    }
+#endif // __x86_64__
+
+    for (auto i = 0u; i < length; i += 8u) {
+        XTEA_encrypt<1>(data + i, k);
+    }
+}
+
+void decrypt(uint8_t* data, size_t length, const key& k)
+{
+    assert(length % 8 == 0);
```

Comment:
this is bad for server availability, if someone throw a unaligned packet to decryption, server will abort

Reply:
It is removed from generated code with `-DCMAKE_BUILD_TYPE=Release` which should definitely be used on production, but I think the server already aligns everything to 8 bytes and the check is redundant.

Reply:
What about the others build type? It should be removed too?

Reply:
If you are debugging your server, you probably want to get assertion errors.